Failing unit-test

### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release


### cbor2 version

5.6.0

### Python version

3.11.5

### What happened?

I ran `python -m pytest -sv tests/` and one of the date tests failed:

```
tests/test_decoder.py:389: in test_date
    assert decoded == expected
E   assert datetime.date(2018, 12, 30) == datetime.date(2018, 12, 31)
```

### How can we reproduce the bug?

run `python -m pytest -sv tests/test_decoder.py`

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -37,6 +37,8 @@ jobs:
         CFLAGS: "-UNDEBUG"
     - name: Test with pytest
       run: coverage run -m pytest -v
+      env:
+        TZ: America/New_York
     - name: Generate coverage report
       run: coverage xml
     - name: Upload Coverage

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -9,6 +9,7 @@ This library adheres to `Semantic Versioning <http://semver.org/>`_.
 
 - Fixed ``__hash__()`` of the C version of the ``CBORTag`` type crashing when there's a recursive
   reference cycle
+- Fixed decoding of epoch-based dates being affected by the local time zone in the C extension
 
 **5.6.1** (2024-02-01)
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -97,6 +97,8 @@ minversion = 4.0.0
 commands = python -m pytest {posargs}
 package = editable
 extras = test
+setenv =
+  TZ=America/New_York
 
 [testenv:docs]
 extras = doc

--- a/source/decoder.c
+++ b/source/decoder.c
@@ -1337,15 +1337,15 @@ static PyObject *
 CBORDecoder_decode_epoch_date(CBORDecoderObject *self)
 {
     // semantic type 100
-    PyObject *num, *tuple, *ret = NULL;
+    PyObject *num, *ordinal, *ret = NULL;
 
     num = decode(self, DECODE_NORMAL);
     if (num) {
         if (PyNumber_Check(num)) {
-            tuple = PyTuple_Pack(1, PyNumber_Multiply(num, PyLong_FromLong(24 * 60 * 60)));
-            if (tuple) {
-                ret = PyDate_FromTimestamp(tuple);
-                Py_DECREF(tuple);
+            ordinal = PyNumber_Add(num, _CBOR2_date_ordinal_offset);
+            if (ordinal) {
+                ret = PyObject_CallMethodObjArgs(PyDateTime_Date, _CBOR2_str_fromordinal, ordinal, NULL);
+                Py_DECREF(ordinal);
             }
         } else {
             PyErr_Format(

--- a/source/encoder.c
+++ b/source/encoder.c
@@ -1056,7 +1056,7 @@ CBOREncoder_encode_date(CBOREncoderObject *self, PyObject *value)
         tmp = PyObject_CallMethodObjArgs(
                 value, _CBOR2_str_toordinal, NULL);
         if (tmp && fp_write(self, "\xD8\x64", 2) == 0) {
-            ret = CBOREncoder_encode_int(self, PyNumber_Subtract(tmp, PyLong_FromLong(719163)));
+            ret = CBOREncoder_encode_int(self, PyNumber_Subtract(tmp, _CBOR2_date_ordinal_offset));
         }
     } else {
         tmp = PyObject_CallMethodObjArgs(

--- a/source/module.c
+++ b/source/module.c
@@ -621,6 +621,7 @@ _CBOR2_init_thread_locals(void)
 
 PyObject *_CBOR2_empty_bytes = NULL;
 PyObject *_CBOR2_empty_str = NULL;
+PyObject *_CBOR2_date_ordinal_offset = NULL;
 PyObject *_CBOR2_str_as_string = NULL;
 PyObject *_CBOR2_str_as_tuple = NULL;
 PyObject *_CBOR2_str_bit_length = NULL;
@@ -638,6 +639,7 @@ PyObject *_CBOR2_str_encode_date = NULL;
 PyObject *_CBOR2_str_Fraction = NULL;
 PyObject *_CBOR2_str_fromtimestamp = NULL;
 PyObject *_CBOR2_str_FrozenDict = NULL;
+PyObject *_CBOR2_str_fromordinal = NULL;
 PyObject *_CBOR2_str_getvalue = NULL;
 PyObject *_CBOR2_str_groups = NULL;
 PyObject *_CBOR2_str_ip_address = NULL;
@@ -971,6 +973,7 @@ PyInit__cbor2(void)
     INTERN_STRING(Fraction);
     INTERN_STRING(fromtimestamp);
     INTERN_STRING(FrozenDict);
+    INTERN_STRING(fromordinal);
     INTERN_STRING(getvalue);
     INTERN_STRING(groups);
     INTERN_STRING(ip_address);
@@ -1000,6 +1003,9 @@ PyInit__cbor2(void)
 
 #undef INTERN_STRING
 
+    if (!_CBOR2_date_ordinal_offset &&
+            !(_CBOR2_date_ordinal_offset = PyLong_FromLong(719163)))
+        goto error;
     if (!_CBOR2_str_utc_suffix &&
             !(_CBOR2_str_utc_suffix = PyUnicode_InternFromString("+00:00")))
         goto error;

--- a/source/module.h
+++ b/source/module.h
@@ -49,6 +49,7 @@ extern PyObject *_CBOR2_str_encode_date;
 extern PyObject *_CBOR2_str_Fraction;
 extern PyObject *_CBOR2_str_fromtimestamp;
 extern PyObject *_CBOR2_str_FrozenDict;
+extern PyObject *_CBOR2_str_fromordinal;
 extern PyObject *_CBOR2_str_getvalue;
 extern PyObject *_CBOR2_str_groups;
 extern PyObject *_CBOR2_str_ip_address;

